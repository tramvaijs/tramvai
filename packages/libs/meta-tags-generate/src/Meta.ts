import identity from '@tinkoff/utils/function/identity';
import isArray from '@tinkoff/utils/is/array';
import has from '@tinkoff/utils/object/has';
import { MetaWalk } from './MetaWalk';
import type { TagRecord, ListSources, TransformValue, Converter } from './Meta.h';

const defaultConverter = (value: TagRecord | string): TagRecord | null =>
  has('tag', value) ? (value as TagRecord) : null;

export class Meta {
  private listSources: ListSources;

  private transformValue: TransformValue;

  private converters: Record<string, Converter>;

  metaWalk: MetaWalk;

  constructor({
    list,
    transformValue = identity,
    converters = {},
    metaWalk,
  }: {
    list: ListSources;
    transformValue?: TransformValue;
    converters?: Record<string, Converter>;
    metaWalk?: MetaWalk;
  }) {
    this.listSources = list;
    this.transformValue = transformValue;
    this.converters = converters;

    this.metaWalk = metaWalk ?? new MetaWalk();
  }

  dataCollection(...args: any[]) {
    this.listSources.map((fn) => fn(this.metaWalk, ...args));

    let result: TagRecord[] = [];

    this.metaWalk.eachMeta((val, key) => {
      if (val.value !== null) {
        const value = this.transformValue(val);
        const converter = this.converters[key];
        const res = (converter || defaultConverter)(value.value);

        if (res) {
          result = result.concat(isArray(res) ? res.filter(identity) : res);
        }
      }
    });

    this.metaWalk.reset();

    return result;
  }
}
