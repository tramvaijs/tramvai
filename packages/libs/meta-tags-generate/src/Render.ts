import reduceObj from '@tinkoff/utils/object/reduce';
import type { Meta } from './Meta';
import { META_DATA_ATTR } from './constants';
import type { TagRecord } from './Meta.h';

const formatAttr = reduceObj((acc, value, key) => `${acc} ${key}="${value}"`, '');

const htmlConverter = ({ tag, attributes, innerHtml }: TagRecord) => {
  const attr = `${formatAttr(attributes)} ${META_DATA_ATTR}="true"`;

  if (innerHtml) {
    return `<${tag}${attr}>${innerHtml}</${tag}>`;
  }

  return `<${tag}${attr}>`;
};

export class Render {
  meta: Meta;

  constructor(meta: Meta) {
    this.meta = meta;
  }

  render(): string {
    return this.meta.dataCollection().map(htmlConverter).join('\n');
  }
}
