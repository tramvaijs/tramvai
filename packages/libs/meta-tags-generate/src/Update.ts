import eachObj from '@tinkoff/utils/object/each';
import type { Meta } from './Meta';
import { META_DATA_ATTR } from './constants';
import type { TagRecord } from './Meta.h';

const create = ({ tag, attributes, innerHtml }: TagRecord) => {
  const newElement = document.createElement(tag);

  if (innerHtml) {
    newElement.innerHTML = innerHtml;
  }

  attributes &&
    eachObj((value, key) => {
      newElement.setAttribute(key, value || '');
    }, attributes);

  newElement.setAttribute(META_DATA_ATTR, 'true');

  return newElement;
};

export class Update {
  meta: Meta;

  constructor(meta: Meta) {
    this.meta = meta;
  }

  update() {
    const head = document.head || document.querySelector('head');
    const addTags = this.meta.dataCollection().map(create);
    const removeTags = head.querySelectorAll(`[${META_DATA_ATTR}]`);

    const fragment = document.createDocumentFragment();

    [].slice.call(removeTags).forEach((tag) => head.removeChild(tag));
    addTags.forEach((tag) => fragment.appendChild(tag));

    head.insertBefore(fragment, head.firstChild);
  }
}
