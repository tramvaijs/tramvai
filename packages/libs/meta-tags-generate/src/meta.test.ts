/**
 * @jest-environment jsdom
 */

import { Meta } from './Meta';
import { Render } from './Render';
import { Update } from './Update';

const converters = {
  title: (value) => ({ tag: 'title', innerHtml: value }),
  ogTitle: (value) => ({ tag: 'meta', attributes: { property: 'og:title', content: value } }),
  canonical: (value) => ({ tag: 'link', attributes: { rel: 'canonical', href: value } }),
};

const generateMeta = (transformValue?) => {
  const defaultMeta = (walker) => {
    walker.updateMeta(0, {
      title: 's',
      ogTitle: 'd',
      ogAlt: { tag: 'meta', attributes: { k: 's' } },
    });
  };
  const configMeta = (walker) => {
    walker.updateMeta(10, { title: 'g', canonical: 'l' });
  };
  return new Meta({ list: [defaultMeta, configMeta], converters, transformValue });
};

describe('meta - integrate tests', () => {
  it('Наполнение данными и преобразование в строку', () => {
    const meta = generateMeta();
    expect(meta.dataCollection()).toMatchSnapshot();

    const render = new Render(meta);
    expect(render.render()).toMatchSnapshot();
  });

  it('Наполнение данными и обновление в dom', () => {
    const meta = generateMeta();
    new Update(meta).update();

    expect(document.head).toMatchSnapshot();
  });

  it('Преобразовываем значения с приоритетом выше дефолтного', () => {
    const meta = generateMeta((item) => {
      if (item.priority > 0) {
        // eslint-disable-next-line no-param-reassign
        item.value = 'transform';
      }
      return item;
    });
    expect(meta.dataCollection()).toMatchSnapshot();

    const render = new Render(meta);
    expect(render.render()).toMatchSnapshot();
  });

  it('Очистка дефолтных meta параметров', () => {
    const defaultMeta = (walker) => {
      walker.updateMeta(0, {
        title: 's',
        ogTitle: 'd',
        canonical: 'b',
        ogAlt: { tag: 'meta', attributes: { k: 's' } },
      });
    };
    const configMeta = (walker) => {
      walker.updateMeta(10, { title: null, canonical: null });
    };
    const meta = new Meta({ list: [defaultMeta, configMeta], converters });

    expect(meta.dataCollection()).toMatchSnapshot();
    const render = new Render(meta);
    expect(render.render()).toMatchSnapshot();

    new Update(meta).update();

    expect(document.head).toMatchSnapshot();
  });

  it('Не перетираем null значения с высоким приоритетом', () => {
    const meta = generateMeta();
    meta.metaWalk.updateMeta(20, { test: null });
    meta.metaWalk.updateMeta(10, { test: 'string' });

    expect(meta.metaWalk.state).toMatchSnapshot();

    const render = new Render(meta);
    expect(render.render()).toMatchSnapshot();
  });

  it('Сериализация и десериализация стейта в metaWalk', () => {
    const meta = generateMeta();

    meta.metaWalk.updateMeta(20, { title: 'hello' });
    const serializableState = meta.metaWalk.getSerializableState();

    expect(serializableState).toEqual([['title', { value: 'hello', priority: 20 }]]);

    const newMeta = generateMeta();

    expect(newMeta.metaWalk.get('title')).toBeUndefined();

    newMeta.metaWalk.mergeValuesFromSerializableState(serializableState);
    expect(newMeta.metaWalk.get('title')).toEqual({ value: 'hello', priority: 20 });
  });
});
