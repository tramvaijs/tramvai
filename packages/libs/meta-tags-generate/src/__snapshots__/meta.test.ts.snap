// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`meta - integrate tests Наполнение данными и обновление в dom 1`] = `
<head>
  <title
    data-meta-dynamic="true"
  >
    g
  </title>
  <meta
    content="d"
    data-meta-dynamic="true"
    property="og:title"
  />
  <meta
    data-meta-dynamic="true"
    k="s"
  />
  <link
    data-meta-dynamic="true"
    href="l"
    rel="canonical"
  />
</head>
`;

exports[`meta - integrate tests Наполнение данными и преобразование в строку 1`] = `
[
  {
    "innerHtml": "g",
    "tag": "title",
  },
  {
    "attributes": {
      "content": "d",
      "property": "og:title",
    },
    "tag": "meta",
  },
  {
    "attributes": {
      "k": "s",
    },
    "tag": "meta",
  },
  {
    "attributes": {
      "href": "l",
      "rel": "canonical",
    },
    "tag": "link",
  },
]
`;

exports[`meta - integrate tests Наполнение данными и преобразование в строку 2`] = `
"<title data-meta-dynamic="true">g</title>
<meta property="og:title" content="d" data-meta-dynamic="true">
<meta k="s" data-meta-dynamic="true">
<link rel="canonical" href="l" data-meta-dynamic="true">"
`;

exports[`meta - integrate tests Не перетираем null значения с высоким приоритетом 1`] = `
Map {
  "test" => {
    "priority": 20,
    "value": null,
  },
}
`;

exports[`meta - integrate tests Не перетираем null значения с высоким приоритетом 2`] = `
"<title data-meta-dynamic="true">g</title>
<meta property="og:title" content="d" data-meta-dynamic="true">
<meta k="s" data-meta-dynamic="true">
<link rel="canonical" href="l" data-meta-dynamic="true">"
`;

exports[`meta - integrate tests Очистка дефолтных meta параметров 1`] = `
[
  {
    "attributes": {
      "content": "d",
      "property": "og:title",
    },
    "tag": "meta",
  },
  {
    "attributes": {
      "k": "s",
    },
    "tag": "meta",
  },
]
`;

exports[`meta - integrate tests Очистка дефолтных meta параметров 2`] = `
"<meta property="og:title" content="d" data-meta-dynamic="true">
<meta k="s" data-meta-dynamic="true">"
`;

exports[`meta - integrate tests Очистка дефолтных meta параметров 3`] = `
<head>
  <meta
    content="d"
    data-meta-dynamic="true"
    property="og:title"
  />
  <meta
    data-meta-dynamic="true"
    k="s"
  />
</head>
`;

exports[`meta - integrate tests Преобразовываем значения с приоритетом выше дефолтного 1`] = `
[
  {
    "innerHtml": "transform",
    "tag": "title",
  },
  {
    "attributes": {
      "content": "d",
      "property": "og:title",
    },
    "tag": "meta",
  },
  {
    "attributes": {
      "k": "s",
    },
    "tag": "meta",
  },
  {
    "attributes": {
      "href": "transform",
      "rel": "canonical",
    },
    "tag": "link",
  },
]
`;

exports[`meta - integrate tests Преобразовываем значения с приоритетом выше дефолтного 2`] = `
"<title data-meta-dynamic="true">transform</title>
<meta property="og:title" content="d" data-meta-dynamic="true">
<meta k="s" data-meta-dynamic="true">
<link rel="canonical" href="transform" data-meta-dynamic="true">"
`;
